mods/classic-taskbar-background-fix.wh.cpp

```diff
@@ -0,0 +1,368 @@
+retry:  //wait until taskbar has properly initialised in order to detect whether current process will become taskbar or not
+    if (isRetry) {
+        if (WaitForSingleObject(g_initThreadStopSignal, 1000) != WAIT_TIMEOUT) {
+            Wh_Log(L"Shutting down InitThreadFunc before success");
+            return FALSE;
+        }
+    }
+    isRetry = true;
+
+    bool abort = false;
+    if (TryInit(&abort, /*callApplyHookOperations*/true)) {
+        return TRUE;  //hooks done
+    }
+    else if (abort) {
+        return FALSE;   //if the taskbar process is already running then subsequent non-taskbar related explorer.exe instances will not be hooked
+    }
+    else {      //taskbar was not yet found, so we need to retry later
+        goto retry;
+    }
+}
+
+void Wh_ModAfterInit(void) {
+    //Run the hook thread only after Wh_ModAfterInit() has been called. This is in order to avoid race condition in calling Wh_ApplyHookOperations().
+    if (g_retryInitInAThread) {
+
+        if (ResumeThread(g_initThread)) {
+            Wh_Log(L"ResumeThread successful");
+            g_retryInitInAThread = false;
+        }
+        else {
+            Wh_Log(L"ResumeThread failed");
+        }
+    }
+    else {  //if the init was done in Wh_ModInit then hooking was done only after that and taskbar repaint was not yet possible until now
+
+        //apply the updated colour immediately
+        TriggerTaskbarRepaint(NULL);
+    }
+}
+
+void LoadSettings() {
+
+    //config option to keep "show desktop" button black
+    //TODO: option to use some other colour, for example the colour of hidden tray icons popup panel background
+    PCWSTR configString = Wh_GetStringSetting(L"RepaintDesktopButton");
+    g_repaintDesktopButton = (wcscmp(configString, L"no") != 0);        //default is yes, so if it is not "no" then it is yes
+    Wh_FreeStringSetting(configString);     
+}
+
+BOOL Wh_ModInit() {
+
+    LoadSettings();
+
+    Wh_Log(L"Init");
+
+
+    bool abort = false;
+    if (TryInit(&abort, /*callApplyHookOperations*/false)) {    //NB! calling Wh_ApplyHookOperations() is not allowed during Wh_ModInit()
+        return TRUE;  //hooks done
+    }
+    else if (abort) {
+        return FALSE;   //if the taskbar process is already running then subsequent non-taskbar related explorer.exe instances will not be hooked
+    }
+    else {      //taskbar was not yet found, maybe it is still initialising in the current process, so we need to retry later
+        g_initThreadStopSignal = CreateEvent(
+            /*lpEventAttributes = */NULL,           // default security attributes
+            /*bManualReset = */TRUE,				// manual-reset event
+            /*bInitialState = */FALSE,              // initial state is nonsignaled
+            /*lpName = */NULL						// object name
+        );
+
+        if (!g_initThreadStopSignal) {
+            Wh_Log(L"CreateEvent failed");
+            return FALSE;
+        }
+
+        g_initThread = CreateThread(

```

Comment:
That's fine, but in my personal opinion this approach is tricky to maintain and easy to get wrong, e.g. look at the complex `Wh_ModUninit` implementation. I'd prefer to just hook `CreateWindowEx` and catch the taskbar creation there. In any case, it's just an FYI/tip/opinion, you can keep it as is.

Reply:
Tried the CreateWindowEx version now. 

It works generally nicely, but there is one bug which is maybe some interaction between Windhawk and ExplorerPatcher. 

The issue manifests if I execute the following sequence: 
* kill explorer.exe
* load the mod
* start explorer.exe 
* unload the mod
Now explorer.exe hangs, tried number of times, happens always for me. 

When I do not kill explorer.exe first, so the mod starts after explorer.exe, then the issue does not manifest during mod unload. Tried number of times as well.

I tested with another mod that uses CreateWindowEx hooking on explorer.exe and same thing happens with that mod as well. This other mod is named classic-taskbar-context-menu and is created by another author. When creating my mod, I did not copy their code, so I do not think that my or their code have same bug. More likely it is something related to hooked function handling either in Windhawk or ExplorerPatcher.

Reply:
Can you post a crash dump, along with that mod version? You can either post a Drive (or similar service) link, or [email it to me](https://ramensoftware.com/contact). I didn't encounter such an issue with my mods, so I suspect there's something you got wrong on your end.

You should be able to get a crash dump with the following steps:
* Open regedit
* Go to: `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\Windows Error Reporting`
* Create a key named `LocalDumps`
* Create a DWORD value named `DumpType` with value `2`
* Trigger the crash
* Go to the `%LocalAppData%\CrashDumps` folder, you should see a dump file in there

> Maybe until this gets resolved, lets use the thread method. And later when no issues with CreateWindowEx manifest anymore, then I can push a new version with CreateWindowEx, which I have already written as well.

Sure, we can do that if we don't find a quick solution or if you don't have the time/desire to work on it, up to you, let me know.